servers/physics/space_sw.cpp

```diff
@@ -37,7 +37,7 @@ _FORCE_INLINE_ static bool _match_object_type_query(CollisionObjectSW *p_object,
 	if (p_object->get_type()==CollisionObjectSW::TYPE_AREA)
 		return p_type_mask&PhysicsDirectSpaceState::TYPE_MASK_AREA;
 
-	if ((p_object->get_layer_mask()&p_layer_mask)==0)
+	if (((p_object->get_layer_mask()&p_layer_mask)==0) && ((p_object->get_collision_mask()&p_layer_mask)==0))
```

Comment:
Hmm, I think this is going to take a bit longer than I expected. Do you know where the 2D layer/mask comparison is done? I feel like this might not be the correct place to put the check, as I've just looked at the equivalent section of code for the 2D physics and it also has similar code to the 3D physics (before I changed it):

	if ((p_object->get_layer_mask()&p_layer_mask)==0)
		return false;

That seems wrong as it's checking A's layer with B's layer, which as you said is incorrect, but it seems to work just fine, so I'm a little confused. Unless that code is just to check both objects are actually assigned to a layer? That would make sense.... in which case I should put the check in the same place the 2D physics does it, so that hopefully I won't have to modify the function arguments (that function doesn't have access to B's collision mask, so I'd have to add it)